spring/spring5/spring5-remote/src/main/java/org/infinispan/spring/remote/session/RemoteApplicationPublishedBridge.java

```diff
@@ -43,25 +48,47 @@ public void unregisterListener() {
    }
 
    @ClientCacheEntryCreated
-   public void processCacheEntryCreated(ClientCacheEntryCreatedEvent event) {
-      taskExecutor.execute(() -> {
-               Session session = (Session) eventSource.get(event.getKey()).get();
-               if (session != null) {
-                  emitSessionCreatedEvent(session);
-               }
-            }
-      );
+   public void processCacheEntryCreated(ClientCacheEntryCustomEvent<byte[]> event) {
+      emitSessionCreatedEvent(readEvent(event).getValue());
    }
 
    @ClientCacheEntryExpired
-   public void processCacheEntryExpired(ClientCacheEntryExpiredEvent event) {
-      springEventsPublisher.ifPresent(p -> p.publishEvent(new SessionExpiredEvent(eventSource, new MapSession((String) event.getKey()))));
+   public void processCacheEntryExpired(ClientCacheEntryCustomEvent<byte[]> event) {
+      emitSessionExpiredEvent(readEvent(event).getValue());
    }
 
    @ClientCacheEntryRemoved
-   public void processCacheEntryDestroyed(ClientCacheEntryRemovedEvent event) {
-      // We create a new session object because there is an incompatibility with the API right now
-      // We should be able to get the value that has been removed to pass it to the event from infinispan
-      springEventsPublisher.ifPresent(p -> p.publishEvent(new SessionDestroyedEvent(eventSource, new MapSession((String) event.getKey()))));
+   public void processCacheEntryDestroyed(ClientCacheEntryCustomEvent<byte[]> event) {
+      emitSessionDestroyedEvent(readEvent(event).getValue());
+   }
+
+   protected KeyValuePair<String, Session> readEvent(ClientCacheEntryCustomEvent<byte[]> event) {
+      byte[] eventData = event.getEventData();
+      ByteBuffer rawData = ByteBuffer.wrap(eventData);
+      byte[] rawKey = readElement(rawData);
+      byte[] rawValue = readElement(rawData);
+      String key = dataFormat.keyToObj(rawKey, whitelist);
+      KeyValuePair keyValuePair;
+      if (rawValue == null) {
+         // This events will hold either an old or a new value almost every time. But there are some corner cases
+         // during rebalance where neither a new or an old value will be present. This if handles this case
+         keyValuePair = new KeyValuePair<>(key, new MapSession(key));
+      } else {
+         keyValuePair = new KeyValuePair<>(key, dataFormat.valueToObj(rawValue, whitelist));
+      }
+      return keyValuePair;
+   }
+
+   private byte[] readElement(ByteBuffer buffer) {
+      byte[] element = null;
+      try {
+         int length = UnsignedNumeric.readUnsignedInt(buffer);
+         element = new byte[length];
+         buffer.get(element);
+      } catch (Exception ex) {
+
```

Comment:
I missed this... we should at least log the error if something went wrong

Reply:
Actually logging is probably best left to `ClientListenerInvocation`... here we should return `null` if `buffer.remaining() == 0`, but otherwise we should assume the buffer contains a full element.

Reply:
I pushed a commit to check the size to your branch @karesti 
Test suite looks good locally but I'll wait for CI to confirm